src/providers/FusionAuthProvider.tsx

```diff
@@ -1,95 +1,138 @@
 
-            return `${baseUrl}/oauth2/${functionType}?${query}`;
+            return `${configuration.baseUrl}/oauth2/${functionType}?${query}`;
         },
-        [baseUrl],
+        [configuration],
     );
 
     const login = useCallback(
         async (state = '') => {
+            const rand = generateRandomString();
+            document.cookie = `lastState=${rand}`;
             const queryParams = {
-                client_id: clientID,
-                scope: scope,
+                client_id: configuration.clientID,
+                scope: configuration.scope,
                 response_type: 'code',
-                redirect_uri: redirectUri,
+                redirect_uri: configuration.redirectUri,
                 code_challenge: await generatePKCE(),
                 code_challenge_method: 'S256',
-                state: `${generateRandomString()}:${state}`,
+                state: `${rand}:${state}`,
             };
             const fullUrl = generateUrl(FunctionType.login, queryParams);
             window.location.assign(fullUrl);
         },
-        [clientID, scope, redirectUri, generateUrl],
+        [configuration, generateUrl],
     );
 
     const logout = useCallback(async () => {
         const queryParams = {
-            client_id: clientID,
-            post_logout_redirect_uri: redirectUri,
-            id_token_hint: idTokenHint ?? '',
+            client_id: configuration.clientID,
+            post_logout_redirect_uri: configuration.redirectUri,
+            id_token_hint: configuration.idTokenHint ?? '',
         };
         const fullUrl = generateUrl(FunctionType.logout, queryParams);
         window.location.assign(fullUrl);
-    }, [clientID, redirectUri, idTokenHint, generateUrl]);
+    }, [configuration, generateUrl]);
 
     const register = useCallback(
         async (state = '') => {
+            const rand = generateRandomString();
+            document.cookie = `lastState=${rand}`;
             const queryParams = {
-                client_id: clientID,
-                scope: scope,
+                client_id: configuration.clientID,
+                scope: configuration.scope,
                 response_type: 'code',
-                redirect_uri: redirectUri,
+                redirect_uri: configuration.redirectUri,
                 code_challenge: await generatePKCE(),
                 code_challenge_method: 'S256',
-                state: `${generateRandomString()}:${state}`,
+                state: `${rand}:${state}`,
             };
             const fullUrl = generateUrl(FunctionType.register, queryParams);
             window.location.assign(fullUrl);
         },
-        [clientID, scope, redirectUri, generateUrl],
+        [configuration, generateUrl],
     );
 
+    useEffect(() => {
+        try {
+            const lastState = document.cookie
+                .split('; ')
+                .find(row => row.startsWith('lastState='))
+                ?.split('=')[1];
```

Comment:
Let's use [js-cookie](https://github.com/js-cookie/js-cookie) to read/write cookies. This is a pretty standard library that makes manipulating cookies a lot cleaner

Reply:
As a follow up you'll need to install the typescript declarations they list at the bottom of the README for it to work most likely